### PR TITLE
Add more regex tests (and fix one atomicity bug)

### PR DESCRIPTION
**Pull request by stephentoub**
One bug fix:
- We allow a positive lookahead to be used to determine whether a loop can be upgraded to be atomic, but we're currently missing the RightToLeft check that distinguishes between positive lookaheads and positive lookbehinds, so we're erroneously making some loops followed by positive lookbehinds atomic when we shouldn't.  Fix that just by ensuring we only traverse Require nodes when they're lookahead rather than behind.

And adding more tests, in part driven by code coverage gaps. Code coverage numbers with this PR:
```
+--------------------------------+--------+--------+--------+
| Module                         | Line   | Branch | Method |
+--------------------------------+--------+--------+--------+
| System.Text.RegularExpressions | 97.82% | 93.84% | 99.01% |
+--------------------------------+--------+--------+--------+
```

Most of the remaining uncovered lines are either things the tool just has trouble seeing (e.g. that nothing after a `ThrowXyz` method is actually reachable) or code that's likely dead but I'm not yet willing to remove out of an abundance of caution.

cc: @danmosemsft, @eerhardt, @pgovind

### CODE REVIEW

File: src/libraries/System.Text.RegularExpressions/tests/Regex.EscapeUnescape.Tests.cs [hunk 1]
```diff
@@ -35,6 +42,12 @@ public void Escape_NullString_ThrowsArgumentNullException()
         public void Unescape(string str, string expected)
         {
             Assert.Equal(expected, Regex.Unescape(str));
+
+            if (expected.Length > 0)
+            {
+                const int Count = 100;
```
Comment (danmosemsft) [inline src/libraries/System.Text.RegularExpressions/tests/Regex.EscapeUnescape.Tests.cs]: I guess the 100 takes it out of some stackalloc path?

File: src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs [hunk 2]
```diff
@@ -462,17 +464,34 @@ public void Match_Timeout()
             Assert.Equal("a", match.Value);
         }
 
-        [Fact]
-        public void Match_Timeout_Throws()
+        [Theory]
+        [InlineData(RegexOptions.None)]
+        [InlineData(RegexOptions.None | (RegexOptions)0x80 /* Debug */)]
```
Comment (danmosemsft) [inline src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs]: Debug just to exercise dump code?

**Review by danmoseley — APPROVED** [review]

**Review by pgovind — APPROVED** [review]
LGTM

File: src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs (hunk 2 above)
Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs]: Yes

File: src/libraries/System.Text.RegularExpressions/tests/Regex.EscapeUnescape.Tests.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/tests/Regex.EscapeUnescape.Tests.cs]: Yes

**Review by eerhardt — APPROVED** [review]
:shipit:

**Pull request merged**